HTTP-63 caching for sync lookups (#94)
getindata/flink-http-connector: README.md

| lookup.partial-cache.expire-after-access | optional | The max time to live for each rows in lookup cache after accessing the entry in the cache. Specify as a [Duration](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/deployment/config/#duration). `lookup.cache` must be set to `PARTIAL` to use this option. See the following <a href="#lookup-cache">Lookup Cache</a> section for more details. |
| lookup.partial-cache.cache-missing-key | optional | This is a boolean that defaults to true. Whether to store an empty value into the cache if the lookup key doesn't match any rows in the table. `lookup.cache` must be set to `PARTIAL` to use this option. See the following <a href="#lookup-cache">Lookup Cache</a> section for more details.                                                                                                                                                    |

> | lookup.max-retries | optional | The max retry times if lookup database failed; default is 3. See the following <a href="#lookup-cache">Lookup Cache</a> section for more details.                                                                                                                                                                                                                                                                                                  |

Reviewer: ```suggestion
| lookup.max-retries | optional | The max retry times if the http lookup failed; default is 3. See the following <a href="#lookup-cache">Lookup Cache</a> section for more details.                                                                                                                                                                                                                                                                                                  |
```
("database" is a left-over from the JDBC connector doc)

| gid.connector.http.sink.writer.request.mode             | optional | Sets Http Sink request submission mode. Two modes are available to select, `single` and `batch` which is the default mode if option is not specified.                                                                                            |
| gid.connector.http.sink.request.batch.size              | optional | Applicable only for `gid.connector.http.sink.writer.request.mode = batch`. Sets number of individual events/requests that will be submitted as one HTTP request by HTTP sink. The default value is 500 which is same as HTTP Sink `maxBatchSize` |

## Lookup Cache 

> The HTTP Client connector can be used in temporal join as a lookup source (aka. dimension table). Currently, only sync lookup mode is supported.

Reviewer 2: What do you mean by " only sync lookup mode is supported"?

Author: > What do you mean by " only sync lookup mode is supported"?

I copied there words from [https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/jdbc/#lookup-cache](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/jdbc/#lookup-cache) as it is the same caching mechanism. I will expand on you points to make it more clear.

By default, lookup cache is not enabled. You can enable it by setting lookup.cache to PARTIAL.

> The lookup cache is used to improve performance of temporal join the HTTP Client connector. By default, lookup cache is not enabled, so all the API requests are sent on the network. When lookup cache is enabled, each process (i.e. TaskManager) will hold a cache. Flink will lookup the cache first, and only send requests 

Reviewer 2: " each process (i.e. TaskManager)" <- and not each instance of source, a.k.a task? (You can have many tasks per one TM).

| gid.connector.http.sink.writer.request.mode             | optional | Sets Http Sink request submission mode. Two modes are available to select, `single` and `batch` which is the default mode if option is not specified.                                                                                            |
| gid.connector.http.sink.request.batch.size              | optional | Applicable only for `gid.connector.http.sink.writer.request.mode = batch`. Sets number of individual events/requests that will be submitted as one HTTP request by HTTP sink. The default value is 500 which is same as HTTP Sink `maxBatchSize` |

## Lookup Cache 

> The HTTP Client connector can be used in temporal join as a lookup source (aka. dimension table).  

Reviewer: (minor) Generally, acronyms such as "aka" aren't recommended in docs.
```suggestion
The HTTP Client connector can be used in temporal join as a lookup source (also known as dimension table).  
```

> By default, lookup cache is not enabled. You can enable it by setting `lookup.cache` to `PARTIAL`. Caching is only enabled if `asyncPolling` = false.

Reviewer: (minor)
```suggestion
By default, the lookup cache is not enabled. You can enable it by setting `lookup.cache` to `PARTIAL`. Caching is only enabled if `asyncPolling` = false.
```

> The scope of the cache is for per job, so long running jobs can benefit from this caching.

Reviewer: (minor)
```suggestion
The scope of the cache is per job, so long running jobs can benefit from this caching.
```

> The lookup cache is used to improve performance of temporal join the HTTP Client connector. By default, lookup cache is not enabled, so all the API requests are sent on the network. When lookup cache is enabled, Flink will look in the cache first, and only send requests 

Reviewer: (minor)
```suggestion
The lookup cache is used to improve the performance of temporal join in the HTTP Client connector. By default, the lookup cache is not enabled, so all the API requests are sent on the network. When the lookup cache is enabled, Flink looks in the cache first, and only sends requests 
```

> on the network when there is no cached value, and update cache with the rows returned. The oldest rows in cache will be expired when the cache hits the max cached rows `lookup.partial-cache.max-rows` or when the row exceeds the max time to live specified by `lookup.partial-cache.expire-after-write` or `lookup.partial-cache.expire-after-access`.

Reviewer: (minor)
```suggestion
on the network when there is no cached value, and updates the cache with the returned rows. The oldest cached rows expire when the cache hits the maximum cached rows `lookup.partial-cache.max-rows` or when the row exceeds the maximum time to live specified by `lookup.partial-cache.expire-after-write` or `lookup.partial-cache.expire-after-access`.
```

> The cached rows might not be the latest, users can tune expiration options to a smaller value to have a better fresh data, but this may increase the number of API requests sent. So this is a balance between throughput and correctness.

Reviewer: (minor)
```suggestion
The cached rows might not be the latest. Users can tune expiration options to a smaller value to have fresher data, but this may increase the number of sent API requests. So this is a balance between throughput and correctness.
```

> A good use of this cache, is when the API responses are very slowly changing; for example master or reference data. 

Reviewer: (minor)
```suggestion
A good use case for enabling the cache is when the API responses are very slowly changing; for example master or reference data. 
```

There are many cases when caching is not appropriate, for example calling an API to get the latest stock price.    

> By default, flink will cache the empty query result for a Primary key, you can toggle the behaviour by setting `lookup.partial-cache.cache-missing-key` to false.

Reviewer: (minor)
```suggestion
By default, Flink caches the empty query result for a Primary key. You can toggle the behaviour by setting `lookup.partial-cache.cache-missing-key` to false.
```
